tests/test_cli.py

```diff
@@ -484,6 +484,92 @@ def test_stack_commands(self, command, success, yes_flag, exit_code):
+            (StackChangeSetStatus.DEFUNCT, False, 1, False),
+            (StackChangeSetStatus.DEFUNCT, True, 1, False),
+            (StackChangeSetStatus.DEFUNCT, False, 1, True),
+        ]
+    )
+    def test_update_with_change_set(self, change_set_status, yes_flag, exit_code, verbose_flag):
+        create_command = self.mock_stack_actions.create_change_set
+        wait_command = self.mock_stack_actions.wait_for_cs_completion
+        execute_command = self.mock_stack_actions.execute_change_set
+        delete_command = self.mock_stack_actions.delete_change_set
+        describe_command = self.mock_stack_actions.describe_change_set
+
+        wait_command.return_value = change_set_status
+
+        response = {
+            "VerboseProperty": "VerboseProperty",
+            "ChangeSetName": "ChangeSetName",
+            "CreationTime": "CreationTime",
+            "ExecutionStatus": "ExecutionStatus",
+            "StackName": "StackName",
+            "Status": "Status",
+            "StatusReason": "StatusReason",
+            "Changes": [
+                {
+                    "ResourceChange": {
+                        "Action": "Action",
+                        "LogicalResourceId": "LogicalResourceId",
+                        "PhysicalResourceId": "PhysicalResourceId",
+                        "Replacement": "Replacement",
+                        "ResourceType": "ResourceType",
+                        "Scope": "Scope",
+                        "VerboseProperty": "VerboseProperty"
+                    }
+                }
+            ]
+        }
+
+        if not verbose_flag:
+            del response["VerboseProperty"]
+            del response["Changes"][0]["ResourceChange"]["VerboseProperty"]
+
+        describe_command.return_value = response
+
+        kwargs = {"args": ["update", "--change-set", "dev/vpc.yaml"]}
+
+        if yes_flag:
+            kwargs["args"].append("-y")
+        else:
+            kwargs["input"] = "y\n"
+
+        if verbose_flag:
+            kwargs["args"].append("-v")
+
+        result = self.runner.invoke(cli, **kwargs)
+
+        change_set_name = create_command.call_args[0][0]
+        assert 'change-set' in change_set_name
+
+        assert wait_command.called_with(change_set_name)
+        assert delete_command.called_with(change_set_name)
+
+        if change_set_status == StackChangeSetStatus.READY:
+            assert execute_command.called_with(change_set_name)
+            assert describe_command.called_with(change_set_name)
+            output = result.output.splitlines()[0]
+            assert yaml.safe_load(output) == response
+
+        if change_set_status == StackChangeSetStatus.DEFUNCT:
+            assert "Failed to create change set" in result.output
+
+        if change_set_status == StackChangeSetStatus.NO_CHANGES:
+            assert "No changes detected" in result.output
+
+        assert result.exit_code == exit_code
```

Comment:
When you have so much conditional logic in a parameterized test, it's better to just make multiple tests, one for each test case you want to test. This is a test trying to be a bunch of different tests at the same time. I think you should break this up so you don't need all this conditional logic in the same test. (Note: I'm fully aware there are a lot of bad examples of unit tests in this file. But we shouldn't make it worse by perpetuating those bad patterns).

Reply:
I have split the tests into 3 test cases.